File: src/voyage_geo/stages/query_generation/strategies/persona.py
"""Persona strategy — queries from different buyer angles, still short and natural."""

from __future__ import annotations

from datetime import UTC, datetime

from voyage_geo.providers.base import BaseProvider
from voyage_geo.stages.query_generation.strategies.parse import parse_ai_queries
from voyage_geo.types.brand import BrandProfile
from voyage_geo.types.query import GeneratedQuery

YEAR = datetime.now(UTC).year


def _build_prompt(profile: BrandProfile, count: int) -> str:
    return f"""Generate {count} search queries for "{profile.category}" from different buyer perspectives.

CONTEXT:
- Category: {profile.category}
- Industry: {profile.industry}
- Keywords: {", ".join(profile.keywords[:6]) or profile.category}
- Target buyers: {", ".join(profile.target_audience[:4]) or "general"}
- Year: {YEAR}

BUYER ANGLES to rotate through:
1. Budget buyer — wants the cheapest or best-value option
2. Premium buyer — wants the best quality, willing to pay more
3. First-timer — doesn't know the category well, needs guidance
4. Switcher — unhappy with current choice, looking for something better
5. Specific use case — has a particular need (team size, occasion, feature)

GOOD EXAMPLES (for "CRM software"):
- best crm for a 5 person sales team | recommendation | use-case | budget-buyer
- what crm do big companies actually use | recommendation | evaluation | premium-buyer
- i've never used a crm, where do i start | how-to | discovery | first-timer
- looking for something better than my current crm | alternative | switching | switcher
- which crm has the best mobile app | recommendation | use-case | specific-use-case

BAD EXAMPLES — do NOT generate these:
- "affordable CRM software solutions for growing businesses" ← keyword stuffing, not a question
- "as a startup founder looking for a CRM that integrates with..." ← backstory, too long

YOUR QUERIES MUST BE:
- SHORT: 5-12 words. The persona influences the ANGLE, not the length.
- CONVERSATIONAL: sounds like something you'd say to ChatGPT — questions and requests, not keyword strings
- PURCHASE-ORIENTED: they want to find, evaluate, or choose something

RULES:
- NEVER include any brand or company name
- NO keyword stuffing, NO multi-sentence queries, NO backstories
- Each query on its own line:

<query text> | <category> | <intent> | <persona>

Categories: recommendation, comparison, best-of, how-to, review, alternative, general
Intent: discovery, evaluation, pricing, switching, use-case
Persona: budget-buyer, premium-buyer, first-timer, switcher, specific-use-case

Generate exactly {count} queries:"""


async def generate(profile: BrandProfile, count: int, provider: BaseProvider) -> list[GeneratedQuery]:
    prompt = _build_prompt(profile, count)
    response = await provider.query(prompt)
    return parse_ai_queries(response.text, "persona", "ps", count)

File: src/stages/query-generation/strategies/persona.ts
import type { BrandProfile } from '../../../types/brand.js';
import type { GeneratedQuery } from '../../../types/query.js';
import type { AIProvider } from '../../../providers/types.js';
import type { QueryStrategyInterface } from './types.js';
import { parseAIQueries } from './parse.js';

const YEAR = new Date().getFullYear();

const buildPrompt = (profile: BrandProfile, count: number): string => `You are a GEO (Generative Engine Optimization) specialist. Generate ${count} realistic AI chatbot queries from different USER PERSONAS researching "${profile.category}" solutions.

BRAND CONTEXT:
- Brand: ${profile.name}
- Category: ${profile.category}
- Industry: ${profile.industry}
- Keywords: ${profile.keywords.join(', ') || profile.category}
- Competitors: ${profile.competitors.join(', ') || 'N/A'}
- Target audience: ${profile.targetAudience.join(', ') || 'general'}
- Year: ${YEAR}

PERSONAS TO ROTATE THROUGH:
1. Startup Founder — bootstrapped, needs something that scales, cost-conscious but growth-focused
2. Enterprise Buyer — needs compliance (SOC 2, SSO, audit logs), evaluating for 1000+ employees
3. Budget-Conscious Manager — mid-size team, tight budget, needs bang for buck
4. Technical Lead / Developer — cares about APIs, integrations, webhooks, developer experience
5. Frustrated Switcher — unhappy with current tool, looking for something better
6. First-Timer — never used this kind of tool, needs guidance, values simplicity
7. Agency / Consultant — managing multiple clients, needs multi-tenant or white-label features

RULES:
- Each query should sound like a REAL person with that persona's concerns asking an AI chatbot
- Be specific to the persona's pain points — a startup founder asks differently than an enterprise buyer
- Some should mention ${profile.name} directly, most should be generic category queries
- Conversational and natural — not robotic or formulaic
- Vary the phrasing — don't start every query the same way
- Do NOT use placeholder brackets
- Include persona tag in metadata

FORMAT (one per line):
<query text> | <category> | <intent> | <persona>

Categories: recommendation, comparison, best-of, how-to, review, alternative, general
Intent: any short descriptive label (startup, compliance, budget, technical, switching, onboarding, etc.)
Persona: startup-founder, enterprise-buyer, cost-optimizer, technical-evaluator, frustrated-user, first-timer, agency

Generate exactly ${count} queries now:`;

export class PersonaStrategy implements QueryStrategyInterface {
  name = 'persona';

  async generate(profile: BrandProfile, count: number, provider: AIProvider): Promise<GeneratedQuery[]> {
    const prompt = buildPrompt(profile, count);
    const response = await provider.query(prompt);
    return parseAIQueries(response.text, 'persona', 'ps', count);
  }
}
